### PR TITLE
Containerized Volume Client Drivers - Design Proposal 

### PR DESCRIPTION
**Pull request by childsb**
This document lists user stories and design for containerizing the hosts volume driver stack.  

Summary : 

"As a system administrator I want to use Persistent Volumes without installing any additional software on each host."

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/22216)

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-storage @saad-ali @thockin

**Comment by childsb** [comment]
Alternative Proposal from Storage SIG-

Volume "attacher" is long running daemon on each host which handles the mount operations for kubelet.  Long running daemon facilitated through daemon sets: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/daemons.md

I will update the design doc with the alternative approach (user stories remain the same)

File: docs/proposals/volume-client-container.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
+# Containerized Volume Clients for Kubernetes
+This document presents user stories and design for containerized volume clients.
+
+# Motivation
+Many storage systems require specific software pakcages to connect.  The current volume architecture requires that volume client software (VCS) is installed on each host for kubelet to access the volumes.  Its a common mistake to forget to install packages on one or more hosts.  And on some platforms the volume client configuration is difficult or impossible.
+
+**Benefits:**
+
+1. No client installed on host.
+2. No client configuration on host.
```
Comment (lvlv) [inline docs/proposals/volume-client-container.md]: Less library dependencies in kubelet, multi client version support.

**Comment by rootfs** [comment]
a flow of using DaemonSet to mount a filesystem is proposed [here](https://gist.github.com/rootfs/3ddb3e32215420d58fb1)

**Comment by jfrazelle** [comment]
This could probably be closed for the same reasons as mentioned here: https://github.com/kubernetes/kubernetes/issues/18333#issuecomment-230067880. Also It's missing a CLA, needs a rebase, and hasn't been updated since March 17.

**Comment by luxas** [comment]
also relevant: https://github.com/kubernetes/kubernetes/issues/26093, https://github.com/kubernetes/kubernetes/pull/20698

I guess someone from @kubernetes/sig-storage should carry this given no cla and no activity

**Comment by wattsteve** [comment]
I think we need to keep this alive right? @rootfs @childsb

**Comment by matchstick** [comment]
@jingxu97 Does this work overlap with any of your containerized storage client work?
@saad-ali

**Comment by saad-ali** [comment]
This is very much the problem that we are trying to solve now for GCI and this captures the problem we want to solve. The details of how this will be implemented need to be worked out.

**Comment by childsb** [comment]
@jingxu97 we would like to bring this in line with the GKE containerized mount.  is there a design doc for that work somewhere?

We are scheming on containerized mount support for the other volume types and would like to begin work soon.

**Comment by luxas** [comment]
This proposal has to move to the community repo

**Comment by jingxu97** [comment]
@childsb I will work on the doc and share with you asap. Please let me know if you have any question meanwhile. Thanks!

**Comment by thockin** [comment]
Closing for now - the new driver work will help here

**Comment by owenhaynes** [comment]
Which ticket replaces this?

**Pull request closed without merging**